_posts/2023-01-06-Algorithms-Graphs-BFS-DFS.md

```diff
@@ -0,0 +1,340 @@
+# Graph traversal algorithms: BFS & DFS
+
+### Intro
+
+Graphs for beginners seem really hard, but the truth is that they are not so difficult when you learn the basics and you really understand the main mechanisms. Let's start with some basic knowledge.
+
+### What is a graph?
+
+- Consists of nodes which are connected by edges
+
+- Each edge can have a direction (directed graph) or not (undirected graph) 
+
+- Additionally, graph can be weighted or not. Weighted means that edge has some value. It can for example represent the distance between two nodes. 
+
+- Example of the graph you can see here (Undirected Cyclic Graph):
+
+<p align="center">
+<img src="https://raw.githubusercontent.com/swiftingio/blog/%2357-Algorithms-DFS-BFS/Graph.svg " width="643">
+</p>
+  
+### Graph representation
+
+Typically we represent a graph by adjacency list (is the most popular option). Every node has a list pointing to another nodes to which has a connection. For an implementation point of view we can use for that a dicionary.
+
+For example, we can construct a adjacancy list like this:
+
+```swift
+var adjList: [Int: [Int]] = [:]
+
+adjList[1] = [2, 3]
+adjList[2] = [1, 4]
+adjList[3] = [1, 4]
+adjList[2] = [5]
+
+```
+
+The second option to represent a graph is a matrix:
+
+| 0    | 1    | 1    | 0    | 0    |
+| ---- | ---- | ---- | ---- | ---- |
+| 1    | 0    | 0    | 0    | 1    |
+| 1    | 0    | 0    | 1    | 0    |
+| 0    | 1    | 1    | 1    | 0    |
+| 0    | 1    | 0    | 0    | 0    |
+
+that shows that element node 1 (columns 1) is connected with node 2 and 3 (so that's why row 2 and row 4 has ones in column 1).
```

Comment:
```suggestion
This matrix shows that element node 1 (columns 1) is connected with node 2 and 3 (so that's why row 2 and row 4 has ones in column 1).
```